File: JAVA/GestionRelationEmployees/src/main/java/models/Request.java
package models;

import java.sql.Timestamp;

public class Request {

    // Valeurs possibles pour status
    public enum Status   { pending, approved, rejected, cancelled }

    // Valeurs possibles pour priority
    public enum Priority { low, medium, high }

    private int       id;
    private int       userId;
    private int       requestTypeId;
    private String    title;
    private String    description;
    private String attachmentUrl;
    private Status    status;
    private Priority  priority;
    private Timestamp submittedDate;
    private Integer   reviewedBy;       // peut être null
    private Timestamp reviewedDate;     // peut être null
    private String    reviewComment;    // peut être null
    private Timestamp createdAt;
    private Timestamp updatedAt;

    // Champ transient pour affichage (nom du type de demande)
    private String requestTypeName;
    // Champ transient pour affichage (nom du reviewer)
    private String reviewerName;
    private String employeeName;

    // ─── Constructeurs ───────────────────────────────────────────────

    public Request() {}

    // Constructeur pour INSERT (champs obligatoires)
    public Request(int userId, int requestTypeId, String title,
                   String description, Priority priority) {
        this.userId        = userId;
        this.requestTypeId = requestTypeId;
        this.title         = title;
        this.description   = description;
        this.priority      = priority;
        this.status        = Status.pending; // par défaut
    }

    // ─── Getters & Setters ───────────────────────────────────────────

    public int getId()                          { return id; }
    public void setId(int id)                   { this.id = id; }

    public int getUserId()                      { return userId; }
    public void setUserId(int userId)           { this.userId = userId; }

    public int getRequestTypeId()               { return requestTypeId; }
    public void setRequestTypeId(int id)        { this.requestTypeId = id; }

    public String getTitle()                    { return title; }
    public void setTitle(String title)          { this.title = title; }

    public String getDescription()              { return description; }
    public void setDescription(String desc)     { this.description = desc; }

    public String getAttachmentUrl() {
        return attachmentUrl;
    }

    public void setAttachmentUrl(String attachmentUrl) {
        this.attachmentUrl = attachmentUrl;
    }

    public Status getStatus()                   { return status; }
    public void setStatus(Status status)        { this.status = status; }

    public Priority getPriority()               { return priority; }
    public void setPriority(Priority priority)  { this.priority = priority; }

    public Timestamp getSubmittedDate()                  { return submittedDate; }
    public void setSubmittedDate(Timestamp submittedDate){ this.submittedDate = submittedDate; }

    public Integer getReviewedBy()                       { return reviewedBy; }
    public void setReviewedBy(Integer reviewedBy)        { this.reviewedBy = reviewedBy; }

    public Timestamp getReviewedDate()                   { return reviewedDate; }
    public void setReviewedDate(Timestamp reviewedDate)  { this.reviewedDate = reviewedDate; }

    public String getReviewComment()                     { return reviewComment; }
    public void setReviewComment(String reviewComment)   { this.reviewComment = reviewComment; }

    public Timestamp getCreatedAt()                      { return createdAt; }
    public void setCreatedAt(Timestamp createdAt)        { this.createdAt = createdAt; }

    public Timestamp getUpdatedAt()                      { return updatedAt; }
    public void setUpdatedAt(Timestamp updatedAt)        { this.updatedAt = updatedAt; }

    // Champs transients (pour l'affichage)
    public String getRequestTypeName()                       { return requestTypeName; }
    public void setRequestTypeName(String requestTypeName)   { this.requestTypeName = requestTypeName; }

    public String getReviewerName()                          { return reviewerName; }
    public void setReviewerName(String reviewerName)         { this.reviewerName = reviewerName; }
    public String getEmployeeName()                          { return employeeName; }
    public void setEmployeeName(String employeeName)         { this.employeeName = employeeName; }


    // ─── toString ────────────────────────────────────────────────────

    @Override
    public String toString() {
        return "Request{id=" + id + ", title='" + title + "', status=" + status + "}";
    }
}

File: src/Entity/Relation/Request.php
<?php

namespace App\Entity\Relation;

use App\Repository\Relation\RequestRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: RequestRepository::class)]
#[ORM\Table(name: 'requests')]
#[ORM\Index(name: 'idx_requests_status', columns: ['status'])]
#[ORM\Index(name: 'idx_requests_priority', columns: ['priority'])]
#[ORM\Index(name: 'idx_requests_submitted_date', columns: ['submitted_date'])]
#[ORM\HasLifecycleCallbacks]
class Request
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(name: 'user_id', type: Types::INTEGER)]
    private ?int $userId = null;

    #[ORM\Column(name: 'request_type_id', type: Types::INTEGER)]
    private ?int $requestTypeId = null;

    #[ORM\Column(length: 255)]
    private ?string $title = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $description = null;

    #[ORM\Column(name: 'attachment_url', length: 500, nullable: true)]
    private ?string $attachmentUrl = null;

    #[ORM\Column(length: 30)]
    private ?string $status = null;

    #[ORM\Column(length: 20)]
    private ?string $priority = null;

    #[ORM\Column(name: 'submitted_date', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $submittedDate = null;

    #[ORM\Column(name: 'reviewed_by', type: Types::INTEGER, nullable: true)]
    private ?int $reviewedBy = null;

    #[ORM\Column(name: 'reviewed_date', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $reviewedDate = null;

    #[ORM\Column(name: 'review_comment', type: Types::TEXT, nullable: true)]
    private ?string $reviewComment = null;

    #[ORM\Column(name: 'created_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\Column(name: 'updated_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $updatedAt = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getUserId(): ?int
    {
        return $this->userId;
    }

    public function setUserId(int $userId): static
    {
        $this->userId = $userId;
        return $this;
    }

    public function getRequestTypeId(): ?int
    {
        return $this->requestTypeId;
    }

    public function setRequestTypeId(int $requestTypeId): static
    {
        $this->requestTypeId = $requestTypeId;
        return $this;
    }

    public function getTitle(): ?string
    {
        return $this->title;
    }

    public function setTitle(string $title): static
    {
        $this->title = $title;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getAttachmentUrl(): ?string
    {
        return $this->attachmentUrl;
    }

    public function setAttachmentUrl(?string $attachmentUrl): static
    {
        $this->attachmentUrl = $attachmentUrl;
        return $this;
    }

    public function getStatus(): ?string
    {
        return $this->status;
    }

    public function setStatus(string $status): static
    {
        $this->status = $status;
        return $this;
    }

    public function getPriority(): ?string
    {
        return $this->priority;
    }

    public function setPriority(string $priority): static
    {
        $this->priority = $priority;
        return $this;
    }

    public function getSubmittedDate(): ?\DateTimeInterface
    {
        return $this->submittedDate;
    }

    public function setSubmittedDate(?\DateTimeInterface $submittedDate): static
    {
        $this->submittedDate = $submittedDate;
        return $this;
    }

    public function getReviewedBy(): ?int
    {
        return $this->reviewedBy;
    }

    public function setReviewedBy(?int $reviewedBy): static
    {
        $this->reviewedBy = $reviewedBy;
        return $this;
    }

    public function getReviewedDate(): ?\DateTimeInterface
    {
        return $this->reviewedDate;
    }

    public function setReviewedDate(?\DateTimeInterface $reviewedDate): static
    {
        $this->reviewedDate = $reviewedDate;
        return $this;
    }

    public function getReviewComment(): ?string
    {
        return $this->reviewComment;
    }

    public function setReviewComment(?string $reviewComment): static
    {
        $this->reviewComment = $reviewComment;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?\DateTimeInterface
    {
        return $this->updatedAt;
    }

    #[ORM\PrePersist]
    public function onPrePersist(): void
    {
        $now = new \DateTime();
        $this->createdAt = $now;
        $this->updatedAt = $now;
    }

    #[ORM\PreUpdate]
    public function onPreUpdate(): void
    {
        $this->updatedAt = new \DateTime();
    }
}

